storybook/sectionsdecoratormodel.cpp

```diff
@@ -0,0 +1,179 @@
+#include "sectionsdecoratormodel.h"
+
+#include <QScopeGuard>
+
+SectionsDecoratorModel::SectionsDecoratorModel(QObject *parent)
+    : QAbstractItemModel{parent}
+{
+}
+
+void SectionsDecoratorModel::setSourceModel(QAbstractItemModel *sourceModel)
+{
+    if (m_sourceModel != nullptr) {
+        qWarning("Changing source model is not supported!");
+        return;
+    }
+
+    m_sourceModel = sourceModel;
+
+    initialize();
+
+    connect(sourceModel, &QAbstractItemModel::modelReset, this, &SectionsDecoratorModel::initialize);
+    connect(sourceModel, &QAbstractItemModel::rowsInserted, this, &SectionsDecoratorModel::initialize);
+    connect(sourceModel, &QAbstractItemModel::rowsRemoved, this, &SectionsDecoratorModel::initialize);
+    connect(sourceModel, &QAbstractItemModel::rowsMoved, this, &SectionsDecoratorModel::initialize);
+
+    emit sourceModelChanged();
+}
+
+QAbstractItemModel* SectionsDecoratorModel::sourceModel() const
+{
+    return m_sourceModel;
+}
+
+QModelIndex SectionsDecoratorModel::index(int row, int column, const QModelIndex &parent) const
+{
+    if (row >= (sourceModel()->rowCount() + row) || column != 0)
+        return QModelIndex();
+
+    return createIndex(row, column);
+}
+
+QModelIndex SectionsDecoratorModel::parent(const QModelIndex &index) const
+{
+    return {};
+}
+
+int SectionsDecoratorModel::rowCount(const QModelIndex &parent) const
+{
+    return m_rowsMetadata.size();
+}
+
+int SectionsDecoratorModel::columnCount(const QModelIndex &parent) const
+{
+    return 1;
+}
+
+QVariant SectionsDecoratorModel::data(const QModelIndex &index, int role) const
+{
+    if(!index.isValid()) return {};
+
+    const int row = index.row();
+    const RowMetadata &rowMetadata = m_rowsMetadata[row];
+
+    if (role == IsFoldedRole) {
+        return rowMetadata.folded;
+    } else if (role == SubitemsCountRole) {
+        return rowMetadata.count;
+    } else if (role == IsSectionRole) {
+        return rowMetadata.isSection;
+    } else if (role == m_sectionRole && rowMetadata.isSection) {
+        return rowMetadata.sectionName;
+    }
+
+    if (!rowMetadata.isSection) {
+        return sourceModel()->data(
+                    sourceModel()->index(row - rowMetadata.offset, 0), role);
+    } else {
+        return QVariant();
+    }
+}
+
+QHash<int, QByteArray> SectionsDecoratorModel::roleNames() const
+{
+    auto roles = m_sourceModel ? m_sourceModel->roleNames() : QHash<int, QByteArray>{};
+    roles.insert(IsSectionRole, QStringLiteral("isSection").toUtf8());
+    roles.insert(IsFoldedRole, QStringLiteral("isFolded").toUtf8());
+    roles.insert(SubitemsCountRole, QStringLiteral("subitemsCount").toUtf8());
```

Comment:
Make the whole `roles` static and initialize it using `std::initializer_list` CTOR

Reply:
I see, yeah it doesn't make much sense since you're extending the role names